server/src/main/java/org/elasticsearch/common/collect/ImmutableOpenMap.java

```diff
@@ -376,6 +378,26 @@ public VType put(KType key, VType value) {
             return mutableMap.put(key, value);
         }
 
+        public VType putIfAbsent(KType key, Supplier<VType> value) {
+            maybeCloneMap();
+            VType present = mutableMap.get(key);
+            if (present == null) {
+                present = value.get();
+                mutableMap.put(key, present);
+            }
+            return present;
+        }
+
+        @SuppressWarnings("unchecked")
+        public void transformValues(UnaryOperator<VType> transformer) {
+            maybeCloneMap();
+            for (int i = 0; i < mutableMap.values.length; i++) {
+                if (mutableMap.values[i] != null) {
+                    mutableMap.values[i] = transformer.apply((VType) mutableMap.values[i]);
+                }
+            }
+        }
```

Comment:
I am not familiar with ImmutableOpenMap and ObjectObjectHashMap. So I am not sure whether we'd prefer these new methods especially the one that access and assign directly to mutableMap.values. I think it is likely more efficient since it avoids allocate another map. But to play on the safe side, I'd just use a regular map in Metadata#Builder#build and then create an ImmutableOpenMap from it with something like the follows:

```
// use a HashMap `m` to populate aliases

var mb = ImmutableOpenMap.<String, Set<Index>>builder(m.size());
m.forEach((k, v) -> mb.put(k, Collections.unmodifiableSet(v)));
...
```

It has one more map allocation. But that feels acceptable and we don't need to touch anything in ImmutableOpenMap.

Reply:
The main reason for using `ImmutableOpenMap` and extending it with new operations was to avoid copying map when building. Example above effectively copies map in `.forEach` call in the end. In contrast ImmutableOpenMap builder reuses internal structure: https://github.com/elastic/elasticsearch/blob/a59c182f9f7e9d1bf3d6eecbc0e44f24ff91d053/server/src/main/java/org/elasticsearch/common/collect/ImmutableOpenMap.java#L350-L352

Overall I do not have a strong preference for either of approaches. Both of them would result in less copying than we have today.

Reply:
Yeah your change is definitely more efficient. It does not copy but still has re-assignment after the transform. So it may not be far off. I was not sure what was the best pratice with hppc maps. I vaguely remembered it being necessary for performance. So let's not use a regular HashMap. 

I google'd a bit and it seems [the official site](http://labs.carrotsearch.com/hppc-api-and-code-examples.html) actually promotes direct buffer access as the fastest approach. But it also checks an `allocated` field since not every element in `values` is assigned. But the website is outdated and there is no long such field. Their GitHub [examples](https://github.com/carrotsearch/hppc/blob/0.8.1/hppc-examples/src/test/java/com/carrotsearch/hppc/examples/HppcExample_004_IteratingOverMaps.java) no longer has such usages but instead uses the iterator which has the following code at its heart:

```java
    protected ObjectObjectCursor<KType, VType> fetch() {
      if (slot < max) {
        KType existing;
        for (slot++; slot < max; slot++) {
          if (!((existing = (KType) keys[slot]) == null)) {
            cursor.index = slot;
            cursor.key = existing;
            cursor.value = (VType) values[slot];
            return cursor;
          }
        }
      }

      if (slot == max && hasEmptyKey) {
        cursor.index = slot;
        cursor.key = null;
        cursor.value = (VType) values[max];
        slot++;
        return cursor;
      }

      return done();
    }
```

I am not entirely sure whether we should use it since its performance is likely worse than directly manipulating the arrays. But it feels safer. Or alternatively we should add the `null` check like how the iterator does it in your version.